feat(legacy-stats): broadcast swap status for privacy coins

mariocynicys opened the pull request:
# This PR now includes a DB migration. Backup your DB before testing!

This PR allows broadcasting the swap status for privacy coins after hiding the identifying information from it (namely: `my_persistent_pub`).

We also decided to clear out `MakerSwapData.taker_pubkey` and `TakerSwapData.maker_pubkey` so not to reveal the other side's persistent pubkey. They can reveal it themselves when they want.

This PR also correct the stats db index pubkeys, where in the past we used to store one persistent pubkey for side X and htlc pubkey for side Y, but now we store persistent pubkey for side X and also the persistent pubkey for side Y. (X & Y here being the maker or taker interchangeably. such a behaviour mainly depended on who's status messages reached the stats node first)

mariocynicys committed:
broadcast swap status for privacy coins

after hiding identifying info from it

mariocynicys committed:
also hide maker_pubkey and taker_pubkey from the status

each is hidden from the other parties side

these values also exist in the negotation messages but they seem to be mostly set to zeros (added a fixme for discussing that)

shamardy commented:
@mariocynicys can you change the target branch to staging?

mariocynicys commented on mm2src/mm2_main/src/lp_swap/saved_swap.rs:
```diff
@@ -143,13 +143,25 @@ impl SavedSwap {
+                        // persistent pubkey as well as the other party's pubkey.
+                        if data.p2p_privkey.is_some() {
+                            data.my_persistent_pub = Default::default();
+                            data.taker_pubkey = Default::default();
```
uuugh...
if the maker isn't privacy but taker is, the maker still won't hide the taker's pubkey :/

we gotta check the taker coin and based on that hide or not hide the taker pubkey

mariocynicys replied:
resolved by [0863b31](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2648/commits/0863b31afe9339debe6c2a78738790debfd9e4f4), which won't let any side (e.g. maker) disclose the persistent pubkey for the other side (e.g. `data.taker_pubkey`, which is the persistent pubkey of the taker).

this means the stats node can't depend on `data.(t/m)aker_pubkey`, but rather should collect `my_persistent_pubkey` from each status message, one of them will correspond to the taker and one will correspond to the maker.

mariocynicys committed:
always remove the other party persistent pubkey from swap status

the other party will send their status anyway to the stats node, and their pubkey will be present in `my_persistent_pub` if they every choose to include it

mariocynicys committed:
combine other fields from both swap parties

aside from the taker and maker gui and version, we now also grap the maker pubkey from the maker and the taker pubkey from the taker. (i.e. we won't rely on the taker to get the maker's pubkey). on the downside, if the maker has a modded kdf that doesn't send the status, then we won't get the maker's pubkey (which is their right anyway)

mariocynicys commented:
@smk762 could u test if u get consistent pubkeys in this one? (the pubkeys that will be stored are the `my_persistent_pubkey`s from each side rather than the `htlc` pubkeys)

mariocynicys committed:
add a migration for the stats db to fix bad pubkeys

this migration replaces the htlc pubkeys with persistent pubkeys with the help of swap jsons. if a specific swap json can not be found, no action is performed and the pubkey stays as is (might be right or might be wrong)

mariocynicys commented on mm2src/mm2_main/src/database/stats_swaps.rs:
```diff
@@ -116,6 +116,40 @@ pub async fn create_and_fill_stats_swaps_from_json_statements(ctx: &MmArc) -> Ve
+                result.push((UPDATE_MAKER_PUBKEY, vec![maker_pubkey, maker_swap.uuid.to_string()]));
+            },
+            Err(e) => {
+                error!("Error {} on getting maker_pubkey for swap {}", e, maker_swap.uuid);
```
cc @shamardy 

i am actually for zeroing out the pubkeys if we can't find them in the swap json. i think that's better than having a pubkey that might be the correct persistent pub or not with a 50% chance.

shamardy replied:
I agree, Why not make them `NULL` though?

mariocynicys replied:
oh yeah that's better

mariocynicys replied:
[3cd3d8c](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2648/commits/3cd3d8c3eaa14dcdd3f3116e54c1e94161481b8c) & [c617a73](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2648/commits/c617a7328bf00625d7e433da88e1c7412f821afc)

onur-ozkan replied:
It would be great to use covered error logs to panic on tests.

shamardy replied:
Just realised a problem, I don't think seednodes keep all the historical records (maker / taker files), will this be a problem for this migration? Maybe we shouldn't do such migration in this case.

shamardy replied:
nevermind, just realized that we only load and migrate the ones we have files for.

shamardy commented on mm2src/mm2_main/src/database/stats_swaps.rs:
```diff
@@ -121,6 +116,40 @@ pub async fn create_and_fill_stats_swaps_from_json_statements(ctx: &MmArc) -> Ve
+    let maker_swaps = SavedSwap::load_all_from_maker_stats_db(ctx).await.unwrap_or_default();
+    let taker_swaps = SavedSwap::load_all_from_taker_stats_db(ctx).await.unwrap_or_default();
+
+    let mut result = vec![(CREATE_STATS_SWAPS_TABLE, vec![])];
```
In relation to https://github.com/KomodoPlatform/komodo-defi-framework/pull/2648#discussion_r2434809619 I don't understand why `CREATE_STATS_SWAPS_TABLE` statement is needed, wouldn't you also need the below among others?
https://github.com/KomodoPlatform/komodo-defi-framework/blob/61474e8767b55ddb8a17e0cf6d5ed9aa15484bd0/mm2src/mm2_main/src/database/stats_swaps.rs#L80-L83

mariocynicys replied:
that's a mistake. i thought we need to return at least one statement if there happens to be no updates needed. but looks like we don't.

mariocynicys replied:
[3cd3d8c](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2648/commits/3cd3d8c3eaa14dcdd3f3116e54c1e94161481b8c)

mariocynicys committed:
review(shamardy): remove CREATE_STATS_SWAPS_TABLE statement

mariocynicys committed:
fix: insert NULL properly as SqlValue::Null

this changes the migration system to use parameters of type SqlValue instead of plain Strings

onur-ozkan commented on mm2src/mm2_main/src/database/my_swaps.rs:
```diff
@@ -4,7 +4,7 @@
 use common::log::debug;
 use common::PagingOptions;
-use db_common::sqlite::rusqlite::{Connection, Error as SqlError, Result as SqlResult, ToSql};
+use db_common::sqlite::rusqlite::{types::Value as SqlValue, Connection, Error as SqlError, Result as SqlResult, ToSql};
```
Nit: Maybe, instead of adding the alias everywhere this is used, re-export it as `SqlValue` from the database root and use it from there?

mariocynicys replied:
[543b672](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2648/commits/543b6723bab9a55324d2f9c8ca818fc97441d285)

onur-ozkan commented on mm2src/mm2_main/src/database/stats_swaps.rs:
```diff
@@ -43,14 +43,9 @@ const INSERT_STATS_SWAP: &str = "INSERT INTO stats_swaps (
     maker_amount,
     taker_amount,
-    is_success,
-    maker_coin_usd_price,
-    taker_coin_usd_price,
```
Q: What happened to these fields? I couldn't see any explanation to this in the PR description.

mariocynicys replied:
the commit msg here might give some context: https://github.com/KomodoPlatform/komodo-defi-framework/pull/2648/commits/22c4652056023be8b83efbb30aff5fccaf192abc

but in essence, the insert statement is split into two parts, a mandatory part (fields that must be set by any status whether maker or taker), like `[m|t]aker_coin`, `[m|t]aker_coin_usd_price`, `[m|t]aker_pubkey`, ... fields. and optional fields like `maker_gui` (e.g. taker doesn't even have access to such value thus can not set it in their side of the status).

this change makes some mandatory fields optional so that we have them set by the maker, taker, or both.

- the pubkeys are now optional because the maker now doens't relay the taker's pubkey in their status for privacy reasons (and vice versa). if the taker wanna disclose their persistent pubkey, they do it themselves via the status message they send.
- the usd_price are also optionalized here because the maker might not be lucky (network issues or etc...) polling the prices from their configured endpoint, which means they will set it to None. but if the taker happens to have polled the prices just fine, then we should use the taker's polled prices instead of the maker's Nones.

onur-ozkan commented on mm2src/mm2_main/src/database/stats_swaps.rs:
```diff
@@ -121,6 +116,54 @@ pub async fn create_and_fill_stats_swaps_from_json_statements(ctx: &MmArc) -> Ve
+                ]));
+            },
+            Err(e) => {
+                error!("Error {} on getting taker_pubkey for swap {}", e, taker_swap.uuid);
```
Same here.

mariocynicys replied:
[644a129](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2648/commits/644a129976be6ec93e55fa9e085c4c64b12ae61a)

onur-ozkan commented on mm2src/mm2_main/src/database/stats_swaps.rs:
```diff
@@ -318,10 +317,79 @@ fn insert_stats_taker_swap_sql_init(swap: &TakerSavedSwap) -> Option<(&'static s
+                params.push((":maker_pubkey", maker_pubkey.into()));
+            },
+            Err(e) => {
+                error!("[{}] Error on getting maker_pubkey for stats: {}", swap.uuid(), e);
```
Let's use `covered_error` here and below.

mariocynicys replied:
[644a129](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2648/commits/644a129976be6ec93e55fa9e085c4c64b12ae61a)

onur-ozkan commented on mm2src/mm2_main/src/lp_swap/maker_swap.rs:
```diff
@@ -2223,10 +2205,8 @@ pub async fn run_maker_swap(swap: RunMakerSwapInput, ctx: MmArc) {
-                                error!("!broadcast_my_swap_status({}): {}", uuid, e);
-                            }
+                        if let Err(e) = broadcast_my_swap_status(&ctx, uuid).await {
+                            error!("!broadcast_my_swap_status({}): {}", uuid, e);
```
Also here.

mariocynicys replied:
[644a129](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2648/commits/644a129976be6ec93e55fa9e085c4c64b12ae61a)

onur-ozkan commented on mm2src/mm2_main/src/lp_swap/saved_swap.rs:
```diff
@@ -143,13 +144,27 @@ impl SavedSwap {
                 if let Some(ref mut event) = swap.events.first_mut() {
                     if let MakerSwapEvent::Started(ref mut data) = event.event {
                         data.secret = H256Json::default();
+                        // Remove the taker's pubkey. Let the stats node get it from the taker directly.
+                        data.taker_pubkey = Default::default();
```
Why not using optional value and setting it to `None` here?

mariocynicys replied:
i wish.
that's for BC since this is part of a p2p-network sent message.

onur-ozkan replied:
We do really need version handling mechanism. This is a really bad workaround :(

onur-ozkan commented on mm2src/mm2_main/src/lp_swap/taker_swap.rs:
```diff
@@ -524,10 +506,8 @@ pub async fn run_taker_swap(swap: RunTakerSwapInput, ctx: MmArc) {
-                                error!("!broadcast_my_swap_status({}): {}", uuid_str, e);
-                            }
+                        if let Err(e) = broadcast_my_swap_status(&ctx, running_swap.uuid).await {
+                            error!("!broadcast_my_swap_status({}): {}", uuid_str, e);
```
Here as well.

mariocynicys replied:
[644a129](https://github.com/KomodoPlatform/komodo-defi-framework/pull/2648/commits/644a129976be6ec93e55fa9e085c4c64b12ae61a)

shamardy commented on mm2src/mm2_main/src/database/stats_swaps.rs:
```diff
@@ -318,10 +317,79 @@ fn insert_stats_taker_swap_sql_init(swap: &TakerSavedSwap) -> Option<(&'static s
+    let mut extra_args = Vec::new();
+    let mut params = OwnedSqlNamedParams::new();
+
+    if let Some(maker_pubkey) = swap.maker_pubkey() {
+        match maker_pubkey {
+            Ok(maker_pubkey) => {
+                extra_args.push("maker_pubkey = :maker_pubkey");
+                params.push((":maker_pubkey", maker_pubkey.into()));
+            },
+            Err(e) => {
+                error!("[{}] Error on getting maker_pubkey for stats: {}", swap.uuid(), e);
+            },
+        }
+    }
+    if let Some(taker_pubkey) = swap.taker_pubkey() {
+        match taker_pubkey {
+            Ok(taker_pubkey) => {
+                extra_args.push("taker_pubkey = :taker_pubkey");
+                params.push((":taker_pubkey", taker_pubkey.into()));
+            },
+            Err(e) => {
+                error!("[{}] Error on getting taker_pubkey for stats: {}", swap.uuid(), e);
+            },
+        }
+    }
```
nit: should we check for cleared pubkeys here and elsewhere and write them as null instead?

mariocynicys replied:
we can, but i wanted them as zeros to distinguish between pubkeys that are hidden vs pubkeys that aren't sent all together.

shamardy replied:
I guess @smk762 is the main user of this feature so we can ask for his opinion

shamardy replied:
Just asked him, we can keep it as is / all zeroes

smk762 commented:
Tested a swap local kdf (as maker) and seednode  running this version.

**On seednode: **
```
sqlite> select * from stats_swaps where uuid = '3808b964-9acb-4e0d-9d25-8a4325b59005';
id|maker_coin|taker_coin|uuid|started_at|finished_at|maker_amount|taker_amount|is_success|maker_coin_ticker|maker_coin_platform|taker_coin_ticker|taker_coin_platform|maker_coin_usd_price|taker_coin_usd_price|maker_pubkey|taker_pubkey|maker_gui|taker_gui|maker_version|taker_version
3454|ARRR|KMD|3808b964-9acb-4e0d-9d25-8a4325b59005|1760947786|1760950264|0.05|0.55|1|ARRR||KMD||0.3931083965|0.0366863638|000000000000000000000000000000000000000000000000000000000000000000||mm2_777||2.5.2-beta_c617a73|
```
 - `SWAPS/STATS/MAKER/3808b964-9acb-4e0d-9d25-8a4325b59005.json` has 
```
          "my_persistent_pub": "000000000000000000000000000000000000000000000000000000000000000000",
          "maker_coin_htlc_pubkey": "02bef0f3557873e359fcd858a80d1ab0cccd3b133de425f4edb2a69a9b96b74579",
          "taker_coin_htlc_pubkey": "027fe9594b1bf50cb0b34087d3cdb1b6ce1c351a050562daee87e5698b21f51b7f",
          "taker_pubkey": "000000000000000000000000000000000000000000000000000000000000000000",
          "maker_coin_htlc_pubkey": "02120da1a42310ba95d6b69d24f2af074d89409c226ee7c82a82d45f4c74cadd21",
          "taker_coin_htlc_pubkey": "02ce5ac2b3e1ccf1ea3adc973ef687acb23fa46321ba6b836dbdb27b753b6947b8"
```


**On local maker KDF:**
```
sqlite> select * from stats_swaps where uuid = '3808b964-9acb-4e0d-9d25-8a4325b59005' ;
id|maker_coin|taker_coin|uuid|started_at|finished_at|maker_amount|taker_amount|is_success|maker_coin_ticker|maker_coin_platform|taker_coin_ticker|taker_coin_platform|maker_coin_usd_price|taker_coin_usd_price|maker_pubkey|taker_pubkey|maker_gui|taker_gui|maker_version|taker_version
19|ARRR|KMD|3808b964-9acb-4e0d-9d25-8a4325b59005|1760947786|1760950264|0.05|0.55|1|ARRR||KMD||0.3931083965|0.0366863638|000000000000000000000000000000000000000000000000000000000000000000||mm2_777||2.5.2-beta_c617a73|
```

 - `SWAPS/STATS/MAKER/3808b964-9acb-4e0d-9d25-8a4325b59005.json` has 
```
smk@jarrah:~/Documents/.kdf/4874fc00e88b26b414ac4cc0d6e0b49d6d7ffc24$ cat ./SWAPS/STATS/MAKER/3808b964-9acb-4e0d-9d25-8a4325b59005.json | jq | grep pub
          "my_persistent_pub": "000000000000000000000000000000000000000000000000000000000000000000",
          "maker_coin_htlc_pubkey": "02bef0f3557873e359fcd858a80d1ab0cccd3b133de425f4edb2a69a9b96b74579",
          "taker_coin_htlc_pubkey": "027fe9594b1bf50cb0b34087d3cdb1b6ce1c351a050562daee87e5698b21f51b7f",
          "taker_pubkey": "000000000000000000000000000000000000000000000000000000000000000000",
          "maker_coin_htlc_pubkey": "02120da1a42310ba95d6b69d24f2af074d89409c226ee7c82a82d45f4c74cadd21",
          "taker_coin_htlc_pubkey": "02ce5ac2b3e1ccf1ea3adc973ef687acb23fa46321ba6b836dbdb27b753b6947b8"
```

 - `SWAPS/MY/3808b964-9acb-4e0d-9d25-8a4325b59005.json` has 
 ```
 smk@jarrah:~/Documents/.kdf/4874fc00e88b26b414ac4cc0d6e0b49d6d7ffc24$ cat ./SWAPS/MY/3808b964-9acb-4e0d-9d25-8a4325b59005.json | jq | grep pub
          "my_persistent_pub": "0371792f7a6846a0da28f3422501927ae103355c02750bc4c4d8430375329a09ac",
          "maker_coin_htlc_pubkey": "02bef0f3557873e359fcd858a80d1ab0cccd3b133de425f4edb2a69a9b96b74579",
          "taker_coin_htlc_pubkey": "027fe9594b1bf50cb0b34087d3cdb1b6ce1c351a050562daee87e5698b21f51b7f",
          "taker_pubkey": "000000000000000000000000000000000000000000000000000000000000000000",
          "maker_coin_htlc_pubkey": "02120da1a42310ba95d6b69d24f2af074d89409c226ee7c82a82d45f4c74cadd21",
          "taker_coin_htlc_pubkey": "02ce5ac2b3e1ccf1ea3adc973ef687acb23fa46321ba6b836dbdb27b753b6947b8"
```

`my_persistent_pub` matches output from `get_public_key`.

mariocynicys commented:
@smk762 the thing is, for privacy coins, the persistent pubkey for the party who is running the privacy coin will not be disclosed (will be set to zeros), for their privacy.

the taker's persistent pubkey here also wasn't set (was NULL) because it seems that the taker didn't send their status (could you please confirm by checking whether `SWAPS/STATS/TAKER/3808b964-9acb-4e0d-9d25-8a4325b59005.json` exists). As I see that taker version and gui is also not set.

If the taker json exists, then could u please post the logs of this stats node.

edit:

> Tested a swap local kdf (as maker) and seednode running this version.

if you mean here that the taker is running an old version, then this taker defiantly didn't send the swap status. because before this PR, if one of the coins in the swap (the maker here using ARRR) is a privacy coin, we don't send any status for stats.

smk762 commented:
Confirmed taker was on older version of KDF. I'll setup a pair on the latest commit of this branch for confirmation testing, and post here once complete.

smk762 commented:
Under maker/taker, there appears to be a different value for `my_persistent_pub` between the json file under `SWAPS/MY` compared to `SWAPS/STATS/{M,T}AKER`. 

**UUID:** `e6851748-8fdb-42db-928e-f9fcdd6e5d20`
**Taker:** `03f394a5c245230e9d2ee3d3fef89b2398e91621f008b6ba9256aaf98fd55c5d46`
**Maker:** `0371792f7a6846a0da28f3422501927ae103355c02750bc4c4d8430375329a09ac`
**KDF commit:** `2.5.2-beta_c617a73` (both sides & seednode)

**seednode**

```
$ cat ./.kdf/DB/a8768e7ff55c6c5041bf79d06b74aeed1bb7aa91/SWAPS/STATS/TAKER/e6851748-8fdb-42db-928e-f9fcdd6e5d20.json | jq | grep pub`
          "my_persistent_pub": "03f394a5c245230e9d2ee3d3fef89b2398e91621f008b6ba9256aaf98fd55c5d46",
          "maker_coin_htlc_pubkey": "03823cc600a347fad1e3c76f45c3780dc7b427a7d8b8aef93aacde870395d59df0",
          "taker_coin_htlc_pubkey": "03f394a5c245230e9d2ee3d3fef89b2398e91621f008b6ba9256aaf98fd55c5d46",
          "maker_pubkey": "000000000000000000000000000000000000000000000000000000000000000000",
          "maker_coin_htlc_pubkey": "0208ea037983db248ff91573efc449595f183a9fe8310cfb1e157647b2c65ea0d6",
          "taker_coin_htlc_pubkey": "027fe9594b1bf50cb0b34087d3cdb1b6ce1c351a050562daee87e5698b21f51b7f"

cat ./.kdf/DB/a8768e7ff55c6c5041bf79d06b74aeed1bb7aa91/SWAPS/STATS/MAKER/e6851748-8fdb-42db-928e-f9fcdd6e5d20.json | jq | grep pub
          "my_persistent_pub": "000000000000000000000000000000000000000000000000000000000000000000",
          "maker_coin_htlc_pubkey": "0208ea037983db248ff91573efc449595f183a9fe8310cfb1e157647b2c65ea0d6",
          "taker_coin_htlc_pubkey": "027fe9594b1bf50cb0b34087d3cdb1b6ce1c351a050562daee87e5698b21f51b7f",
          "taker_pubkey": "000000000000000000000000000000000000000000000000000000000000000000",
          "maker_coin_htlc_pubkey": "03823cc600a347fad1e3c76f45c3780dc7b427a7d8b8aef93aacde870395d59df0",
          "taker_coin_htlc_pubkey": "03f394a5c245230e9d2ee3d3fef89b2398e91621f008b6ba9256aaf98fd55c5d46"
```

`sqlite> select * from stats_swaps where uuid = 'e6851748-8fdb-42db-928e-f9fcdd6e5d20';`
```
id|maker_coin|taker_coin|uuid|started_at|finished_at|maker_amount|taker_amount|is_success|maker_coin_ticker|maker_coin_platform|taker_coin_ticker|taker_coin_platform|maker_coin_usd_price|taker_coin_usd_price|maker_pubkey|taker_pubkey|maker_gui|taker_gui|maker_version|taker_version
3483|ARRR|KMD|e6851748-8fdb-42db-928e-f9fcdd6e5d20|1760961162|1760963489|0.0181818181818182|0.2|1|ARRR||KMD||0.396987|0.03699919|000000000000000000000000000000000000000000000000000000000000000000|03f394a5c245230e9d2ee3d3fef89b2398e91621f008b6ba9256aaf98fd55c5d46|mm2_777|MM2CLI|2.5.2-beta_c617a73|2.5.2-beta_c617a73
```

**MAKER:**

```
$ cat ./SWAPS/STATS/MAKER/e6851748-8fdb-42db-928e-f9fcdd6e5d20.json | jq | grep pub
          "my_persistent_pub": "000000000000000000000000000000000000000000000000000000000000000000",
          "maker_coin_htlc_pubkey": "0208ea037983db248ff91573efc449595f183a9fe8310cfb1e157647b2c65ea0d6",
          "taker_coin_htlc_pubkey": "027fe9594b1bf50cb0b34087d3cdb1b6ce1c351a050562daee87e5698b21f51b7f",
          "taker_pubkey": "000000000000000000000000000000000000000000000000000000000000000000",
          "maker_coin_htlc_pubkey": "03823cc600a347fad1e3c76f45c3780dc7b427a7d8b8aef93aacde870395d59df0",
          "taker_coin_htlc_pubkey": "03f394a5c245230e9d2ee3d3fef89b2398e91621f008b6ba9256aaf98fd55c5d46"

$ cat ./SWAPS/STATS/TAKER/e6851748-8fdb-42db-928e-f9fcdd6e5d20.json | jq | grep pub
          "my_persistent_pub": "03f394a5c245230e9d2ee3d3fef89b2398e91621f008b6ba9256aaf98fd55c5d46",
          "maker_coin_htlc_pubkey": "03823cc600a347fad1e3c76f45c3780dc7b427a7d8b8aef93aacde870395d59df0",
          "taker_coin_htlc_pubkey": "03f394a5c245230e9d2ee3d3fef89b2398e91621f008b6ba9256aaf98fd55c5d46",
          "maker_pubkey": "000000000000000000000000000000000000000000000000000000000000000000",
          "maker_coin_htlc_pubkey": "0208ea037983db248ff91573efc449595f183a9fe8310cfb1e157647b2c65ea0d6",
          "taker_coin_htlc_pubkey": "027fe9594b1bf50cb0b34087d3cdb1b6ce1c351a050562daee87e5698b21f51b7f"

$ cat ./SWAPS/MY/e6851748-8fdb-42db-928e-f9fcdd6e5d20.json | jq | grep pub
          "my_persistent_pub": "0371792f7a6846a0da28f3422501927ae103355c02750bc4c4d8430375329a09ac",
          "maker_coin_htlc_pubkey": "0208ea037983db248ff91573efc449595f183a9fe8310cfb1e157647b2c65ea0d6",
          "taker_coin_htlc_pubkey": "027fe9594b1bf50cb0b34087d3cdb1b6ce1c351a050562daee87e5698b21f51b7f",
          "taker_pubkey": "000000000000000000000000000000000000000000000000000000000000000000",
          "maker_coin_htlc_pubkey": "03823cc600a347fad1e3c76f45c3780dc7b427a7d8b8aef93aacde870395d59df0",
          "taker_coin_htlc_pubkey": "03f394a5c245230e9d2ee3d3fef89b2398e91621f008b6ba9256aaf98fd55c5d46"
```

`sqlite> select * from stats_swaps where uuid = 'e6851748-8fdb-42db-928e-f9fcdd6e5d20';`
```
id|maker_coin|taker_coin|uuid|started_at|finished_at|maker_amount|taker_amount|is_success|maker_coin_ticker|maker_coin_platform|taker_coin_ticker|taker_coin_platform|maker_coin_usd_price|taker_coin_usd_price|maker_pubkey|taker_pubkey|maker_gui|taker_gui|maker_version|taker_version
21|ARRR|KMD|e6851748-8fdb-42db-928e-f9fcdd6e5d20|1760961162|1760963489|0.0181818181818182|0.2|1|ARRR||KMD||0.396987|0.03699919|000000000000000000000000000000000000000000000000000000000000000000|03f394a5c245230e9d2ee3d3fef89b2398e91621f008b6ba9256aaf98fd55c5d46|mm2_777|MM2CLI|2.5.2-beta_c617a73|2.5.2-beta_c617a73
```

**TAKER:**
```
$ cat ./SWAPS/STATS/MAKER/e6851748-8fdb-42db-928e-f9fcdd6e5d20.json | jq | grep pub
          "my_persistent_pub": "000000000000000000000000000000000000000000000000000000000000000000",
          "maker_coin_htlc_pubkey": "0208ea037983db248ff91573efc449595f183a9fe8310cfb1e157647b2c65ea0d6",
          "taker_coin_htlc_pubkey": "027fe9594b1bf50cb0b34087d3cdb1b6ce1c351a050562daee87e5698b21f51b7f",
          "taker_pubkey": "000000000000000000000000000000000000000000000000000000000000000000",
          "maker_coin_htlc_pubkey": "03823cc600a347fad1e3c76f45c3780dc7b427a7d8b8aef93aacde870395d59df0",
          "taker_coin_htlc_pubkey": "03f394a5c245230e9d2ee3d3fef89b2398e91621f008b6ba9256aaf98fd55c5d46"

$ cat ./SWAPS/STATS/TAKER/e6851748-8fdb-42db-928e-f9fcdd6e5d20.json | jq | grep pub
          "my_persistent_pub": "03f394a5c245230e9d2ee3d3fef89b2398e91621f008b6ba9256aaf98fd55c5d46",
          "maker_coin_htlc_pubkey": "03823cc600a347fad1e3c76f45c3780dc7b427a7d8b8aef93aacde870395d59df0",
          "taker_coin_htlc_pubkey": "03f394a5c245230e9d2ee3d3fef89b2398e91621f008b6ba9256aaf98fd55c5d46",
          "maker_pubkey": "000000000000000000000000000000000000000000000000000000000000000000",
          "maker_coin_htlc_pubkey": "0208ea037983db248ff91573efc449595f183a9fe8310cfb1e157647b2c65ea0d6",
          "taker_coin_htlc_pubkey": "027fe9594b1bf50cb0b34087d3cdb1b6ce1c351a050562daee87e5698b21f51b7f"

$ cat ./SWAPS/MY/e6851748-8fdb-42db-928e-f9fcdd6e5d20.json | jq | grep pub
          "my_persistent_pub": "0371792f7a6846a0da28f3422501927ae103355c02750bc4c4d8430375329a09ac",
          "maker_coin_htlc_pubkey": "0208ea037983db248ff91573efc449595f183a9fe8310cfb1e157647b2c65ea0d6",
          "taker_coin_htlc_pubkey": "027fe9594b1bf50cb0b34087d3cdb1b6ce1c351a050562daee87e5698b21f51b7f",
          "taker_pubkey": "000000000000000000000000000000000000000000000000000000000000000000",
          "maker_coin_htlc_pubkey": "03823cc600a347fad1e3c76f45c3780dc7b427a7d8b8aef93aacde870395d59df0",
          "taker_coin_htlc_pubkey": "03f394a5c245230e9d2ee3d3fef89b2398e91621f008b6ba9256aaf98fd55c5d46"
```

`sqlite> select * from stats_swaps where uuid = 'e6851748-8fdb-42db-928e-f9fcdd6e5d20';`
```
id|maker_coin|taker_coin|uuid|started_at|finished_at|maker_amount|taker_amount|is_success|maker_coin_ticker|maker_coin_platform|taker_coin_ticker|taker_coin_platform|maker_coin_usd_price|taker_coin_usd_price|maker_pubkey|taker_pubkey|maker_gui|taker_gui|maker_version|taker_version
21|ARRR|KMD|e6851748-8fdb-42db-928e-f9fcdd6e5d20|1760961162|1760963489|0.0181818181818182|0.2|1|ARRR||KMD||0.396987|0.03699919|000000000000000000000000000000000000000000000000000000000000000000|03f394a5c245230e9d2ee3d3fef89b2398e91621f008b6ba9256aaf98fd55c5d46|mm2_777|MM2CLI|2.5.2-beta_c617a73|2.5.2-beta_c617a73
```

mariocynicys commented:
@smk762 yup that's the expected behavior.

let's not regard `SWAPS/MY/` since it contains full info and comparing it to `SWAPS/STATS/` at it is kinda cheating.

the info saved in stats jsons and stats DB is masked as expected (for privacy coins, persistent pubkeys shouldn't be leaked).

If you wanna test that persistent pubkeys are stored correctly for both taker and maker (in stats DB), then u gotta use two non-privacy coins in the swap so u can view both persistent pubkeys.

the stats this PR provide for the privacy coins exclude the persistent pubkey for the party using the privacy coin, since that defeats the anonymity aspect of using the privacy coin.

mariocynicys committed:
review(onur): use covered error

mariocynicys committed:
review(onur): re-export rusqlite::type::Value as SqlValue from common::database

mariocynicys committed:
resolve a fixme: it is what it is

dunno if recreate_taker_swap even works due to this, but also my_persistent_pub might not be that useful after all for the party who is receiving the recreated swap since they already have hold of that public key already

smk762 reviewed (approved):
Output matches expected, thanks.

onur-ozkan commented on mm2src/common/log.rs:
```diff
@@ -176,6 +176,18 @@ macro_rules! covered_warn {
     };
 }
 
+#[macro_export]
+/// A macro to log errors in production environment and panic in tests.
+macro_rules! covered_error {
+    ($($arg:tt)+) => {
+        if cfg!(not(test)) {
+            common::log::error!($($arg)+)
+        } else {
+            panic!($($arg)+)
+        }
+    };
+}
```
Please rebase on dev to avoid conflict. This already exists on the dev branch.

shamardy replied:
This PR is needed in staging for the 2.5.2-beta release though

mariocynicys replied:
> Please rebase on dev to avoid conflict. This already exists on the dev branch.

yeah i copied it exactly as is from dev. i expect git is smart enough to not complain about this xD
but we can't pull all of dev now though, since some stuff there aren't intended to go into staging yet.

onur-ozkan replied:
It's not an exact copy tho, the doc-comment is misplaced.

shamardy replied:
will fix it when back propagating to dev as a new commit / merge commit

shamardy reviewed (approved):
Thanks a lot for the fix!
